**Chi-Square Test**

Basit olarak 2 kategori cinsindeki değerin arasındaki ilişki olup olmadığına bakıyoruz.

X ve Y iki kategori değeri dersek

H0 -> X ve Y arasında bir bağlantı yok. (Bağımsız - Independent)
HA -> X ve Y bağımlı. (Dependent)

In [59]:
exp_1 = pd.DataFrame(columns = [" ", "Heart_Rate_Increased", "No_Heart_Rate_Increased" ,"Totals"])

In [66]:
exp_1 = pd.DataFrame(columns = [" ", "Heart_Rate_Increased", "No_Heart_Rate_Increased" ,"Totals"])
exp_1[" "] = ["Treated","Not Treated","Total"]
exp_1["Heart_Rate_Increased"] = [36,30,66]
exp_1["No_Heart_Rate_Increased"] = [14,25,39]
exp_1["Totals"] = [50,55,105]
exp_1 = exp_1.set_index(" ")
display(exp_1)

Heart_Rate_Increased  No_Heart_Rate_Increased  Totals
                                                                  
Treated                        36                       14      50
Not Treated                    30                       25      55
Total                          66                       39     105

In [75]:
exp_1[" "] = ["Treated","Not Treated","Total"]
exp_1["Heart_Rate_Increased"] = ["36/50","30/55","66/105"]
exp_1["No_Heart_Rate_Increased"] = ["14/50","25/50","39/105"]
exp_1["Totals"] = ["50/105","55/105","105"]
exp_1 = exp_1.set_index(" ")
display(exp_1)

Heart_Rate_Increased No_Heart_Rate_Increased  Totals
                                                                
Treated                    36/50                   14/50  50/105
Not Treated                30/55                   25/50  55/105
Total                     66/105                  39/105     105

In [65]:
exp_1 = pd.DataFrame(columns = [" ", "Heart_Rate_Increased", "No_Heart_Rate_Increased" ,"Totals"])
exp_1[" "] = ["Treated","Not Treated","Total"]
exp_1["Heart_Rate_Increased"] = [36/50,30/55,66/105]
exp_1["No_Heart_Rate_Increased"] = [14/50,25/50,39/105]
exp_1["Totals"] = [50/105,55/105,105]
exp_1 = exp_1.set_index(" ")
pd.options.display.float_format = '{:.2f}'.format
display(exp_1)

Heart_Rate_Increased  No_Heart_Rate_Increased  Totals
                                                                  
Treated                      0.72                     0.28    0.48
Not Treated                  0.55                     0.50    0.52
Total                        0.63                     0.37  105.00

Örnek olarak: 0.72 oranı Tedavi görmüş bir hayvanın kalp hızının artma oranı  P(Treated, Hearth Rate Increased)<br>
P(Treated) = 0.48

$$ {P(X,Y) = P(X) + P(Y)}$$

Eğer olayların birbirinden bağımsız olduğunu kabul edersek olayların çarpımı birlikte aynı anda olma olasılığı ile aynı olur.

In [68]:
exp_1[" "] = ["Treated","Not Treated","Total"]
exp_1["Heart_Rate_Increased"] = ["0.72 x 0.30","0.28 x 0.18", 66/105]
exp_1["No_Heart_Rate_Increased"] = ["0.55 x 0.33","0.50 x 0.19",39/105]
exp_1["Totals"] = [50/105,55/105,105]
exp_1 = exp_1.set_index(" ")
display(exp_1)

Heart_Rate_Increased No_Heart_Rate_Increased  Totals
                                                                
Treated              0.72 x 0.30             0.55 x 0.33    0.48
Not Treated          0.28 x 0.18             0.50 x 0.19    0.52
Total                       0.63                    0.37  105.00

$$x^2 = \sum \frac{(observed - expected)^2}{expected}$$

In [74]:
result = ((0.72-0.30)**2/0.30)+((0.55-0.33)**2/0.33)+((0.28-0.18)**2/0.18)+((0.50-0.19)**2/0.19)
print("Result = {:.3f}".format(result))

Result = 1.296


**Dof Hesaplama**: Degree of freedom genel olarak değişkenlerin boyutları - 1 <br>
Bizim örneğimizde:<br>
(2-1)x(2-1) = 1 dof

In [5]:
import pandas as pd
import numpy as np

In [41]:

table = pd.DataFrame(columns = ['Df',"0.5","0.1","0.05","0.02","0.01","0.001"])

table["Df"] = [1,2,3,4,5]
table["0.5"] = [0.455,1.386,2.366,3.357,4.351]
table["0.1"] = [2.706,4.605,6.251,7.779,9.236]
table["0.05"] = [3.841,5.991,7.815,9.488,11.07]
table["0.02"] = [5.412,7.824,9.837,11.668,13.388]
table["0.01"] = [6.635,9.21,11.345,13.277,15.086]
table["0.001"] = [10.827,13.815,16.268,18.465,20.517]
table.set_index("Df")

0.5    0.1    0.05    0.02    0.01   0.001
Df                                              
1   0.455  2.706   3.841   5.412   6.635  10.827
2   1.386  4.605   5.991   7.824   9.210  13.815
3   2.366  6.251   7.815   9.837  11.345  16.268
4   3.357  7.779   9.488  11.668  13.277  18.465
5   4.351  9.236  11.070  13.388  15.086  20.517

**x² > table value = accept = dependent(H1)** <br>

**x² ≤ table value = reject = independent(H0)**

**0.05 için 3.841 değerinden sonucumuz(1.296) küçük olduğu için H0 yani iki değer birbirinden bağımsız.**

In [91]:
from scipy import stats


a1 = [36,14]
a2 = [30,25]

chi2_stat, p_val, dof, ex = stats.chi2_contingency([a1,a2])
print("===Chi2 Stat===")
print(chi2_stat)
print("\n")
print("===Degrees of Freedom===")
print(dof)
print("\n")
print("===P-Value===")
print(p_val)
print("\n")
print("===Contingency Table===")
print(ex)

===Chi2 Stat===
2.710942466624286


===Degrees of Freedom===
1


===P-Value===
0.09966209595851808


===Contingency Table===
[[31.42857143 18.57142857]
 [34.57142857 20.42857143]]


**p-value değerimiz = 0.099 > 0.05 H0 kabul ediyoruz ve Independent.**